colexecdisk: fix incorrect context capture in partitionerToOperator

yuzefovich opened the pull request:
This commit fixes recently exposed minor issue with how `partitionerToOperator` captures the context. In particular, these operators are reused multiple times by the external sort and the hash-based partitioner for each new partition, and previously these operator would capture the context that was provided on the first call to `Init`. In case of the external sort, that context comes from the ordered synchronizer which has a tracing span that is finished after each partition, so the captured context would have already finished span that can be problematic down the line.

The fix in this commit is to make the operator resettable so that the new context is captured every time the operator is reused. Calling `Reset` is already in place in the hash-based partitioner (because other operators are reused there too), and this commit adds a call in the external sort.

Fixes: #125177.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125236)
<!-- Reviewable:end -->

DrewKimball reviewed (commented):
> Reviewed 1 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125236)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/sql/colexec/colexecdisk/utils.go` line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125236#-NzigFHBAwd49-ehvQsN:-NzigFHBAwd49-ehvQsO:bxofhr2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cd921f312f4dbd9308d33cfc3656559619f5b3c/pkg/sql/colexec/colexecdisk/utils.go#L40)):*
> ```Go
> //
> // ctx field must be set explicitly by the user before this operator is being
> // used.
> ```

This works, but would it be nicer to make this a resettable operator? That would allow passing in a new context, but still reuse already allocated memory.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
colexecdisk: fix incorrect context capture in partitionerToOperator

This commit fixes recently exposed minor issue with how
`partitionerToOperator` captures the context. In particular, these
operators are reused multiple times by the external sort and the
hash-based partitioner for each new partition, and previously these
operators would capture the context that was provided on the first call
to `Init`. In case of the external sort, that context comes from the
ordered synchronizer which has a tracing span that is finished after
each partition, so the captured context would have already finished span
that can be problematic down the line.

The fix in this commit is to make the operator resettable so that the
new context is captured every time the operator is reused. Calling
`Reset` is already in place in the hash-based partitioner (because other
operators are reused there too), and this commit adds a call in the
external sort.

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125236)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/colexec/colexecdisk/utils.go` line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125236#-NzigFHBAwd49-ehvQsN:-NzjLIIT6ZLe8RiGnKlZ:b-jswti5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cd921f312f4dbd9308d33cfc3656559619f5b3c/pkg/sql/colexec/colexecdisk/utils.go#L40)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

This works, but would it be nicer to make this a resettable operator? That would allow passing in a new context, but still reuse already allocated memory.
</blockquote></details>

Very nice, I forgot about that! Done.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks!

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125236#-:-NzjMQBSDQsNLhPoA0bU:b-g2a5fb)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
